rocketpy/motors/Tank.py

```diff
@@ -0,0 +1,842 @@
+        gases at each part of the tank (body and caps) as well as its masses.
+
+        Parameters
+        ----------
+        t : float
+            Time at which the state of the tank is to be calculated.
+
+        Returns
+        -------
+        None
+        """
+        self.evaluateFilling(t)
+        self.evaluateMassDistribution()
+        self.evaluateCentroids()
+
+    def evaluateFilling(self, t):
+        """Calculates the distribution of volume of liquid and gases in the tank
+        portions at a given time.
+
+        Parameters
+        ----------
+        t : float
+            Time at which the filling of the tank is to be calculated.
+
+        Returns
+        -------
+        None
+        """
+        liquidVolume = self.liquidVolume.getValueOpt(t)
+
+        if 0 <= liquidVolume < self.bottomCap.volume:
+            self.bottomCap.filled_volume = liquidVolume
+            self.cylinder.filled_volume = 0
+            self.upperCap.filled_volume = 0
+        elif 0 < liquidVolume <= self.bottomCap.volume + self.cylinder.volume:
+            self.bottomCap.filled_volume = self.bottomCap.volume
+            self.cylinder.filled_volume = liquidVolume - self.bottomCap.volume
+            self.upperCap.filled_volume = 0
+        elif 0 < liquidVolume <= self.volume:
+            self.bottomCap.filled_volume = self.bottomCap.volume
+            self.cylinder.filled_volume = self.cylinder.volume
+            self.upperCap.filled_volume = liquidVolume - (
+                self.bottomCap.volume + self.cylinder.volume
+            )
+        else:
+            raise ValueError(
+                f"{self.name} tank liquid volume is either negative or greater than "
+                "total tank volume. Check input data to make sure it is correct."
+            )
+
+    def evaluateMassDistribution(self):
+        """Calculates the mass distribution of liquid and gases at the tank based
+        on the volume distribution.
+
+        Returns
+        -------
+        None
+        """
+        self.bottomCapLiquidMass = self.liquid.density * self.bottomCap.filled_volume
+        self.cylinderLiquidMass = self.liquid.density * self.cylinder.filled_volume
+        self.upperCapLiquidMass = self.liquid.density * self.upperCap.filled_volume
+
+        self.bottomCapGasMass = self.gas.density * self.bottomCap.empty_volume
+        self.cylinderGasMass = self.gas.density * self.cylinder.empty_volume
+        self.upperCapGasMass = self.gas.density * self.upperCap.empty_volume
+
+        self.bottomCapMass = self.bottomCapLiquidMass + self.bottomCapGasMass
+        self.cylinderMass = self.cylinderLiquidMass + self.cylinderGasMass
+        self.upperCapMass = self.upperCapLiquidMass + self.upperCapGasMass
+
+    def evaluateCentroids(self):
+        """Calculates the centroids of the liquid and gaseous portions of the tank
+        based on the volume distributions.
```

Comment:
TODO: Make it clear relative to which point are the centroids calculated (coordinate system orientation and origin).